fixedwidth/src/main/java/org/apache/metamodel/fixedwidth/FixedWidthReader.java

```diff
@@ -18,78 +18,235 @@
+     */
+    public String[] readLine() throws IllegalStateException {
         try {
-            line = _reader.readLine();
-            return _parser.parseLine(line);
+            beforeReadLine();
+            _rowNumber++;
+            return getValues();
         } catch (IOException e) {
             throw new IllegalStateException(e);
         }
-	}
-	
+    }
+
+    /**
+     * Empty hook that enables special behavior in sub-classed readers (by overriding this method). 
+     */
+    protected void beforeReadLine() {
+        return;
+    }
+
+    private String[] getValues() throws IOException {
+        final List<String> values = new ArrayList<>();
+        final String singleRecordData = readSingleRecordData();
+
+        if (singleRecordData == null) {
+            return null;
+        }
+
+        processSingleRecordData(singleRecordData, values);
+        String[] result = values.toArray(new String[values.size()]);
+
+        if (!_failOnInconsistentLineWidth && !_constantWidth) {
+            result = correctResult(result);
+        }
+
+        validateConsistentValue(singleRecordData, result, values.size());
+
+        return result;
+    }
+
+    private void validateConsistentValue(String recordData, String[] result, int valuesSize) {
+        if (!_failOnInconsistentLineWidth) {
+            return;
+        }
+
+        InconsistentValueWidthException inconsistentValueException = null;
+
+        if (_constantWidth) {
+            if (recordData.length() % _fixedValueWidth != 0) {
+                inconsistentValueException = new InconsistentValueWidthException(result, recordData, _rowNumber);
+            }
+        } else if (result.length != valuesSize || recordData.length() != _expectedLineLength) {
+            inconsistentValueException = new InconsistentValueWidthException(result, recordData, _rowNumber);
+        }
+
+        if (inconsistentValueException != null) {
+            throw inconsistentValueException;
+        }
+    }
+
+    private void processSingleRecordData(final String singleRecordData, final List<String> values) {
+        StringBuilder nextValue = new StringBuilder();
+        final CharacterIterator it = new StringCharacterIterator(singleRecordData);
+        _valueIndex = 0;
+
+        for (char c = it.first(); c != CharacterIterator.DONE; c = it.next()) {
+            processCharacter(c, nextValue, values, singleRecordData);
+        }
+
+        if (nextValue.length() > 0) {
+            addNewValueIfAppropriate(values, nextValue);
+        }
+    }
+
+    String readSingleRecordData() throws IOException {
+        StringBuilder line = new StringBuilder();
+        int ch;
+
+        for (ch = _stream.read(); !isEndingCharacter(ch); ch = _stream.read()) {
+            line.append((char) ch);
+        }
+
+        if (ch == CARRIAGE_RETURN) {
+            readLineFeedIfFollows();
+        }
+
+        return (line.length()) > 0 ? line.toString() : null;
```

Comment:
Here, the returned String is charset-less which is actually a bug for diacritics input. PR for this bug fix is here: https://github.com/apache/metamodel/pull/121